File: TextosJapones/src/main/java/org/jisho/textosJapones/fileparse/ParseFactory.java
package org.jisho.textosJapones.fileparse;

import org.jisho.textosJapones.util.Util;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.io.File;
import java.io.IOException;

public class ParseFactory {

    private static final Logger LOGGER = LoggerFactory.getLogger(ParseFactory.class);

    public static Parse create(String file) {
        return create(new File(file));
    }

    public static Parse create(File file) {
        Parse parser = null;
        String fileName = file.getAbsolutePath().toLowerCase();

        if (Util.isRar(fileName)) {
            parser = new RarParse();
        } else if (Util.isZip(fileName)) {
            parser = new ZipParse();
        } else if (Util.isSevenZ(fileName)) {
            parser = new SevenZipParse();
        } else if (Util.isTarball(fileName)) {
            parser = new TarParse();
        }
        return tryParse(parser, file);
    }

    private static Parse tryParse(Parse parse, File file) {
        if (parse == null) {
            return null;
        }
        try {
            parse.parse(file);
        } catch (IOException e) {
            LOGGER.error(e.getMessage(), e);
            return null;
        }
        return parse;
    }
}

File: ProcessaTexto/src/main/kotlin/br/com/fenix/processatexto/fileparse/ParseFactory.kt
package br.com.fenix.processatexto.fileparse

import br.com.fenix.processatexto.util.Utils
import org.slf4j.LoggerFactory
import java.io.File
import java.io.IOException
import java.util.*

class ParseFactory {
    companion object {
        private val LOGGER = LoggerFactory.getLogger(ParseFactory::class.java)

        fun create(file: String): Parse {
            return create(File(file))
        }

        fun create(file: File): Parse {
            val parser: Parse
            val fileName = file.absolutePath.lowercase(Locale.getDefault())
            if (Utils.isRar(fileName))
                parser = RarParse()
            else if (Utils.isZip(fileName))
                parser = ZipParse()
            else if (Utils.isSevenZ(fileName))
                parser = SevenZParse()
            else if (Utils.isTarball(fileName))
                parser = TarParse()
            else
                throw Exception("Tipo não implementado")

            return tryParse(parser, file)
        }

        private fun tryParse(parse: Parse, file: File): Parse {
            try {
                parse.parse(file)
            } catch (e: IOException) {
                LOGGER.error(e.message, e)
                throw Exception("Não foi possível abrir o arquivo")
            }
            return parse
        }
    }
}